File: backend/app/routers/stays.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List, Optional
from .. import models, schemas
from ..database import get_db

router = APIRouter(prefix="/stays", tags=["stays"])

@router.get("/", response_model=List[schemas.StayList])
def get_stays(
    location: Optional[str] = None,
    property_type: Optional[str] = None,
    min_price: Optional[float] = None,
    max_price: Optional[float] = None,
    min_capacity: Optional[int] = None,
    db: Session = Depends(get_db)
):
    """
    Get all stays with optional filters:
    - location: Filter by location
    - property_type: cottage, villa, tent, homestay
    - min_price/max_price: Price range per night
    - min_capacity: Minimum guest capacity
    """
    query = db.query(models.Stay)
    
    if location:
        query = query.filter(models.Stay.location.ilike(f"%{location}%"))
    if property_type:
        query = query.filter(models.Stay.property_type == property_type)
    if min_price is not None:
        query = query.filter(models.Stay.price_per_night >= min_price)
    if max_price is not None:
        query = query.filter(models.Stay.price_per_night <= max_price)
    if min_capacity is not None:
        query = query.filter(models.Stay.capacity >= min_capacity)
    
    stays = query.all()
    return stays

@router.get("/{stay_id}", response_model=schemas.Stay)
def get_stay(stay_id: int, db: Session = Depends(get_db)):
    """Get a single stay by ID with full details including amenities"""
    stay = db.query(models.Stay).filter(models.Stay.id == stay_id).first()
    if not stay:
        raise HTTPException(status_code=404, detail="Stay not found")
    return stay

@router.post("/", response_model=schemas.Stay)
def create_stay(stay: schemas.StayCreate, db: Session = Depends(get_db)):
    """Create a new stay (Admin only - auth to be added later)"""
    
    # Create the stay
    db_stay = models.Stay(
        name=stay.name,
        slug=stay.slug,
        property_type=stay.property_type,
        description=stay.description,
        location=stay.location,
        price_per_night=stay.price_per_night,
        capacity=stay.capacity,
        bedrooms=stay.bedrooms,
        bathrooms=stay.bathrooms,
        rating=stay.rating,
        is_featured=stay.is_featured
    )
    
    # Add images
    for img in stay.images:
        db_img = models.StayImage(url=img.url)
        db_stay.images.append(db_img)
    
    # Add amenities
    if stay.amenity_ids:
        amenities = db.query(models.Amenity).filter(models.Amenity.id.in_(stay.amenity_ids)).all()
        db_stay.amenities = amenities
    
    db.add(db_stay)
    db.commit()
    db.refresh(db_stay)
    return db_stay

File: backend-node/src/routes/stays.js
const express = require("express");
const router = express.Router();
const prisma = require("../db");

// GET /stays?location=...&property_type=...&min_price=...&max_price=...&min_capacity=...
router.get("/", async (req, res, next) => {
  try {
    const { location, property_type, min_price, max_price, min_capacity } = req.query;

    const where = { AND: [] };

    if (location) {
      where.AND.push({ location: { contains: location, mode: "insensitive" } });
    }
    if (property_type) {
      where.AND.push({ propertyType: property_type });
    }
    if (min_price !== undefined) {
      where.AND.push({ pricePerNight: { gte: Number(min_price) } });
    }
    if (max_price !== undefined) {
      where.AND.push({ pricePerNight: { lte: Number(max_price) } });
    }
    if (min_capacity !== undefined) {
      where.AND.push({ capacity: { gte: Number(min_capacity) } });
    }

    const stays = await prisma.stay.findMany({
      where: where.AND.length ? where : undefined,
      include: {
        images: true,
        amenities: { include: { amenity: true } },
      },
    });

    // Flatten amenities for response parity with Python
    const result = stays.map((s) => ({
      ...s,
      amenities: s.amenities.map((a) => a.amenity),
    }));

    res.json(result);
  } catch (err) {
    console.error("GET /stays Error:", err);
    next(err);
  }
});

// GET /stays/:id
router.get("/:id", async (req, res, next) => {
  try {
    const stay = await prisma.stay.findUnique({
      where: { id: Number(req.params.id) },
      include: {
        images: true,
        amenities: { include: { amenity: true } },
      },
    });

    if (!stay) return res.status(404).json({ detail: "Stay not found" });

    res.json({
      ...stay,
      amenities: stay.amenities.map((a) => a.amenity),
    });
  } catch (err) {
    next(err);
  }
});

// POST /stays
router.post("/", async (req, res, next) => {
  try {
    const {
      name, slug, propertyType, description, location,
      pricePerNight, capacity, bedrooms, bathrooms,
      rating, isFeatured, images = [], amenityIds = [],
    } = req.body;

    const stay = await prisma.stay.create({
      data: {
        name, slug, propertyType, description, location,
        pricePerNight, capacity,
        bedrooms: bedrooms ?? 1,
        bathrooms: bathrooms ?? 1,
        rating: rating ?? 0.0,
        isFeatured: isFeatured ?? false,
        images: { create: images.map((img) => ({ url: img.url })) },
        amenities: {
          create: amenityIds.map((id) => ({ amenityId: id })),
        },
      },
      include: {
        images: true,
        amenities: { include: { amenity: true } },
      },
    });

    res.status(201).json({
      ...stay,
      amenities: stay.amenities.map((a) => a.amenity),
    });
  } catch (err) {
    next(err);
  }
});

module.exports = router;
